src/utils/updateWeights.ts

```diff
@@ -0,0 +1,60 @@
+import type { NumberArray } from 'cheminfo-types';
+
+import { xAbsolute } from '../x/xAbsolute';
+import { xMedian } from '../x/xMedian';
+import { xSubtract } from '../x/xSubtract';
+
+interface UpdateWeightsOptions {
+  controlPoints?: NumberArray;
+  learningRate: number;
+  minWeight: number;
+  residualFactor: number;
+}
+
+/**
+ * Update the weights based on the residuals between the original data and the new baseline.
+ * @param yData - The original data.
+ * @param newBaseline - The new baseline calculated.
+ * @param weights - The current weights to be updated.
+ * @param options - Options for updating weights.
+ * @param options.learningRate - The learning rate for updating weights.
+ * @param options.minWeight - The minimum weight value.
+ * @param options.residualFactor - The factor to calculate the threshold for residuals.
+ * @param options.controlPoints - Control points to determine which weights to update.
+ * @returns The updated weights.
+ */
+
+export function updateWeights(
+  yData: NumberArray,
+  newBaseline: NumberArray,
+  weights: NumberArray,
+  options: UpdateWeightsOptions,
+) {
+  const { controlPoints, residualFactor, learningRate, minWeight } = options;
+  const absResiduals = xAbsolute(xSubtract(yData, newBaseline));
+
+  const medAbsRes = xMedian(absResiduals);
+  const mad = 1.4826 * medAbsRes;
```

Comment:
What is this magic number? Can you add a comment?
And what is `mad` ? Can you find a better name?

Reply:
You have not answered or addressed this comment.

Reply:
Sorry, I misunderstood with the 3 somewhere, this is a constant to approximately calculate the MAD. Assuming a normal distribution. It is the inverse of the quartile 0.75

Reply:
Please add this as a comment (and explain what MAD is).